File: examples/store/src/store/recommender.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 AimpathyMinds

"""Mock product recommender for the yaagents store example.

Returns same-category products as recommendations.
Replace mock_recommend() with a real LLM call to make this a genuine
AI-powered recommender. See examples/store/skills/ for starter prompts.
"""

from __future__ import annotations

from .data import get_customer, get_product, load_products

def mock_recommend(
    product_id: str,
    customer_id: str | None = None,
    limit: int = 3,
    exclude_purchased: bool = True,
) -> tuple[list[dict], str]:
    """Return same-category recommendations for a product.

    Args:
        product_id: ID of the seed product.
        customer_id: Optional customer ID; used to filter purchased items.
        limit: Maximum number of recommendations to return.
        exclude_purchased: When True and customer_id is set, filter out
            products already purchased by this customer.

    Returns:
        A (recommendations, reasoning) tuple. recommendations is a list
        of product dicts; reasoning is a human-readable explanation string.
    """
    seed = get_product(product_id)
    if not seed:
        return [], "Product not found"

    all_products = load_products()
    candidates = [
        p
        for p in all_products
        if p["category"] == seed["category"] and p["id"] != product_id
    ]

    if customer_id and exclude_purchased:
        customer = get_customer(customer_id)
        if customer:
            purchased = set(customer.get("purchased", []))
            candidates = [p for p in candidates if p["id"] not in purchased]

    recommendations = candidates[:limit]
    reason = f"Same category as '{seed['name']}' ({seed['category']})"
    if customer_id:
        reason += f"; personalised for customer {customer_id} (purchased items excluded)"
    return recommendations, reason

File: examples/store-go/recommender.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 AimpathyMinds

package main

import "fmt"

// mockRecommend returns same-category products as recommendations.
//
// Replace this function with a real LLM call to build a genuine
// AI-powered recommender. See examples/store/skills/ for starter prompts
// (they apply equally to the Python and Go implementations).
func mockRecommend(
	products []Product,
	seed *Product,
	customer *Customer,
	limit int,
	excludePurchased bool,
) ([]Product, string) {
	if limit <= 0 {
		limit = 3
	}

	// Build purchased set for filtering.
	purchased := map[string]bool{}
	if customer != nil && excludePurchased {
		for _, pid := range customer.Purchased {
			purchased[pid] = true
		}
	}

	// Collect same-category candidates.
	var candidates []Product
	for _, p := range products {
		if p.Category == seed.Category && p.ID != seed.ID && !purchased[p.ID] {
			candidates = append(candidates, p)
		}
	}

	// Truncate to limit.
	if len(candidates) > limit {
		candidates = candidates[:limit]
	}

	reason := fmt.Sprintf("Same category as '%s' (%s)", seed.Name, seed.Category)
	if customer != nil {
		reason += fmt.Sprintf("; personalised for customer %s (purchased items excluded)", customer.ID)
	}
	return candidates, reason
}
